stage #2.5/module #4 REST Basics - Spring Web/#6.REST API Tools for Testing, Design, and Development/swagger_for_designing_and_documenting_rest_api.md

```diff
@@ -0,0 +1,201 @@
+
+This file let us:
+
+- Describe all available API endpoints (e.g. /users, /users/{id})
+- Operations in the endpoints such as GET /users, POST /user
+- Input and Output parameters for each operation
+- Authentication mechanisms
+- Contact information, API Licence, terms of use and other information
+
+## Swagger
+Swagger is a set of open-source tools built around the OpenAPI Specification that can help you to design, build, document and consume REST APIs. The ability of APIs to describe their own structure is the root of all awesomeness in Swagger. Note that Swagger not only helps you to design and document REST APIs, it also lets you build (Server Stub) and Consume (Rest Clients) REST APIs.
+
+The major Swagger tools include:
+
+- **Swagger Editor** — browser-based editor where one can write OpenAPI specification
+- **Swagger UI** — renders OpenAPI specs as interactive API documentation
+- **Swagger Codegen** — generates server stubs and client libraries from an OpenAPI specification
+
+Here is a sample OpenAPI document in the Swagger Editor:
+
+![](./media/a_sample_open_api_document_in_the_swagger_editor.png "A sample OpenAPI document in the Swagger Editor")
+
+## Swagger in Action
+
+To enable Swagger in a RESTful application you should add the following dependencies:
+
+> implementation group: 'io.springfox', name: 'springfox-swagger2', version: '3.0.0'
+implementation group: 'io.springfox', name: 'springfox-swagger-ui', version: '3.0.0'
+
+Add Swagger configuration in a separate configuration class:
+
+```Java
+@Configuration
+@EnableSwagger2
+public class SwaggerConfiguration {
+
+    @Bean
+    public Docket api(){
+        return new Docket(DocumentationType.SWAGGER_2)
+                .select()
+                .apis(RequestHandlerSelectors.basePackage("mjc.newsapplication.controller"))
+                .paths(PathSelectors.any())
+                .build().apiInfo(apiEndPointInfo());
+    }
+
+    public ApiInfo apiEndPointInfo(){
+        return new ApiInfoBuilder().title("Application Rest API")
+                .description("News Application API")
+                .contact(new Contact("MJC Supervisor", "mjc/finalexam", "mjc@epam.com"))
+                .license("Apache 2.0")
+                .licenseUrl("http://www.apache.org/licenses/LICENSE-2.0.html")
+                .version("0.0.1-SNAPSHOT")
+                .build();
+    }
+}
+```
+Here, You have added metadata information about the REST API such as API name, author, website, license and so on. You have also instructed Swagger to only generate documentation for the components present in the mjc.newsapplication.controller package.
+
+Since we have enabled Swagger, let us see the documentation of our API endpoints done by Swagger. This is rendered through Swagger UI in the following link:
+
+http://localhost:8080//swagger-ui.html#/news-controller
```

Comment:
duplicate 'slash' symbol after port specified - will cause an error trying to use such URL